File: common/transform.lua
--- @class rt.Transform
rt.Transform = meta.class("Transform")

--- @brief initialize as identity
function rt.Transform:instantiate()
    self._native = love.math.newTransform()
end

--- @brief apply another transform to this one
function rt.Transform:apply(other)
    meta.assert(other, rt.Transform)
    self._native:apply(other._native)
end

--- @brief get a copy of this transform
function rt.Transform:clone()
    local out = rt.Transform()
    out._native:setMatrix(self._native:getMatrix())
    return out
end

--- @brief replace the matrix values directly
function rt.Transform:_update(
    x11, x12, x13, x14,
    x21, x22, x23, x24,
    x31, x32, x33, x34,
    x41, x42, x43, x44
)
    self._native:setMatrix(
        x11, x12, x13, x14,
        x21, x22, x23, x24,
        x31, x32, x33, x34,
        x41, x42, x43, x44
    )
end

--- @brief
function rt.Transform:translate(...)
    if select("#", ...) == 2 then
        self._native:transform(select(1, ...), select(2, ...))
    else
        local x, y, z = select(1, ...), select(2, ...), select(3, ...)

        local x11, x12, x13, x14,
        x21, x22, x23, x24,
        x31, x32, x33, x34,
        x41, x42, x43, x44 = self._native:getMatrix()

        x14 = x14 + x11 * x + x12 * y + x13 * z
        x24 = x24 + x21 * x + x22 * y + x23 * z
        x34 = x34 + x31 * x + x32 * y + x33 * z

        self._native:setMatrix(
            x11, x12, x13, x14,
            x21, x22, x23, x24,
            x31, x32, x33, x34,
            x41, x42, x43, x44
        )
    end
end

--- @brief
function rt.Transform:scale(...)
    if select("#", ...) == 2 then
        self._native:scale(select(1, ...), select(2, ...))
    else
        local x, y, z = select(1, ...), select(2, ...), select(3, ...)

        local x11, x12, x13, x14,
        x21, x22, x23, x24,
        x31, x32, x33, x34,
        x41, x42, x43, x44 = self._native:getMatrix()

        x11 = x11 * x
        x12 = x12 * y
        x13 = x13 * z
        x21 = x21 * x
        x22 = x22 * y
        x23 = x23 * z
        x31 = x31 * x
        x32 = x32 * y
        x33 = x33 * z

        self._native:setMatrix(
            x11, x12, x13, x14,
            x21, x22, x23, x24,
            x31, x32, x33, x34,
            x41, x42, x43, x44
        )
    end
end

--- @brief inverse
function rt.Transform:inverse()
    local out = rt.Transform()
    out._native = self._native:inverse()
    return out
end

--- @brief
function rt.Transform:as_inverse()
    self._native = self._native:inverse()
    return self
end

--- @brief transpose
function rt.Transform:transpose()
    local x11, x12, x13, x14,
    x21, x22, x23, x24,
    x31, x32, x33, x34,
    x41, x42, x43, x44 = self._native:getMatrix()

    self._native:setMatrix(
        x11, x21, x31, x41,
        x12, x22, x32, x42,
        x13, x23, x33, x43,
        x14, x24, x34, x44
    )
end

--- @brief
function rt.Transform:override(
    x11, x12, x13, x14,
    x21, x22, x23, x24,
    x31, x32, x33, x34,
    x41, x42, x43, x44
)
    self._native:setMatrix(
        x11, x12, x13, x14,
        x21, x22, x23, x24,
        x31, x32, x33, x34,
        x41, x42, x43, x44
    )
end

local _reusable = love.math.newTransform()

--- @brief rotate around the X axis
function rt.Transform:rotate_x(angle)
    local c = math.cos(angle)
    local s = math.sin(angle)
    _reusable:setMatrix(
        1, 0, 0, 0,
        0, c, -s, 0,
        0, s, c, 0,
        0, 0, 0, 1
    )
    self._native:apply(_reusable)
end

--- @brief rotate around the Y axis
function rt.Transform:rotate_y(angle)
    local c = math.cos(angle)
    local s = math.sin(angle)
    _reusable:setMatrix(
        c, 0, s, 0,
        0, 1, 0, 0,
        -s, 0, c, 0,
        0, 0, 0, 1
    )
    self._native:apply(_reusable)
end

--- @brief rotate around the Z axis
function rt.Transform:rotate_z(angle)
    local c = math.cos(angle)
    local s = math.sin(angle)
    _reusable:setMatrix(
        c, -s, 0, 0,
        s, c, 0, 0,
        0, 0, 1, 0,
        0, 0, 0, 1
    )
    self._native:apply(_reusable)
end

--- @brief transform a point by this transform
function rt.Transform:transform_point(...)
    if select("#", ...) == 2 then
        return self._native:transformPoint(select(1, ...), select(2, ...))
    else
        local x, y, z = select(1, ...), select(2, ...), select(3, ...)
        local w = 1

        local x11, x12, x13, x14,
        x21, x22, x23, x24,
        x31, x32, x33, x34,
        x41, x42, x43, x44 = self._native:getMatrix()

        local rx = x * x11 + y * x12 + z * x13 + w * x14
        local ry = x * x21 + y * x22 + z * x23 + w * x24
        local rz = x * x31 + y * x32 + z * x33 + w * x34
        local rw = x * x41 + y * x42 + z * x43 + w * x44

        return rx, ry, rz, rw
    end
end

--- @brief transform a point by the inverse of this transform
function rt.Transform:inverse_transform_point(...)
    if select("#", ...) == 2 then
        return self._native:inverseTransformPoint(select(1, ...), select(2, ...))
    else
        local x, y, z = select(1, ...), select(2, ...), select(3, ...)
        local w = 1

        local x11, x12, x13, x14,
        x21, x22, x23, x24,
        x31, x32, x33, x34,
        x41, x42, x43, x44 = self._native:inverse():getMatrix()

        local rx = x * x11 + y * x12 + z * x13 + w * x14
        local ry = x * x21 + y * x22 + z * x23 + w * x24
        local rz = x * x31 + y * x32 + z * x33 + w * x34
        local rw = x * x41 + y * x42 + z * x43 + w * x44

        return rx, ry, rz, rw
    end
end

function rt.Transform:as_perspective_projection(fov, aspect, near, far)
    local top = near * math.tan(fov / 2)
    local bottom = -top
    local right = top * aspect
    local left = -right

    self._native:setMatrix(
        2 * near / (right - left), 0, 0, 0,
        0, 2 * near / (bottom - top), 0, 0,
        0, 0, -1 * (far + near) / (far - near), -2 * far * near / (far - near),
        0, 0, -1, 1
    )

    return self
end

function rt.Transform:as_orthographic_projection(width, height, near, far)
    local aspect = width / height
    local top = -height / 2
    local bottom = -top
    local right = top * -aspect
    local left = -right
    self._native:setMatrix(
        2 / (right - left), 0, 0, -(right + left) / (right - left),
        0, 2 / (top - bottom), 0, -(top + bottom) / (top - bottom),
        0, 0, -2 / (far - near), -(far + near) / (far - near),
        0, 0, 0, 1
    )

    return self
end

--- @brief
function rt.Transform:look_at(
    eye_x, eye_y, eye_z, target_x,
    target_y, target_z,
    up_x, up_y, up_z
)
    local z1, z2, z3 = math.normalize(eye_x - target_x, eye_y - target_y, eye_z - target_z)
    local x1, x2, x3 = math.normalize(math.cross(up_x, up_y, up_z, z1, z2, z3))
    local y1, y2, y3 = math.cross(z1, z2, z3, x1, x2, x3)

    self._native:setMatrix(
        x1, x2, x3, -math.dot(x1, x2, x3, eye_x, eye_y, eye_z),
        y1, y2, y3, -math.dot(y1, y2, y3, eye_x, eye_y, eye_z),
        z1, z2, z3, -math.dot(z1, z2, z3, eye_x, eye_y, eye_z),
        0, 0, 0, 1
    )

    return self
end

--- @brief
function rt.Transform:set_target_to(
    eye_x, eye_y, eye_z,
    target_x, target_y, target_z,
    up_x, up_y, up_z
)
    local z1, z2, z3 = math.normalize(eye_x - target_x, eye_y - target_y, eye_z - target_z)
    local x1, x2, x3 = math.normalize(math.cross(up_x, up_y, up_z, z1, z2, z3))
    local y1, y2, y3 = math.cross(z1, z2, z3, x1, x2, x3)

    self._native:setMatrix(
        x1, y1, z1, eye_x,
        x2, y2, z2, eye_y,
        x3, y3, z3, eye_z,
        0, 0, 0, 1
    )
end

--- @brief
function rt.Transform:set_orientation(
    forward_x, forward_y, forward_z,
    right_x, right_y, right_z,
    up_x, up_y, up_z
)
    local _, _, _, x14,
    _, _, _, x24,
    _, _, _, x34,
    x41, x42, x43, x44 = self._native:getMatrix()

    forward_x, forward_y, forward_z = math.normalize(forward_x, forward_y, forward_z)
    right_x, right_y, right_z = math.normalize(right_x, right_y, right_z)
    up_x, up_y, up_z = math.normalize(up_x, up_y, up_z)

    self._native:setMatrix(
        right_x, right_y, right_z, x14,
        up_x, up_y, up_z, x24,
        forward_x, forward_y, forward_z, x34,
        x41, x42, x43, x44
    )
end

--- @brief Set the translation component of the transform
function rt.Transform:set_position(x, y, z)
    local x11, x12, x13, _,
    x21, x22, x23, _,
    x31, x32, x33, _,
    x41, x42, x43, x44 = self._native:getMatrix()

    self._native:setMatrix(
        x11, x12, x13, x,
        x21, x22, x23, y,
        x31, x32, x33, z,
        x41, x42, x43, x44
    )
end

--- @brief Set orientation from a quaternion (i, j, k, w)
function rt.Transform:set_to_orientation_from_quaternion(i, j, k, w)
    self._native:setMatrix(
        1 - 2 * j * j - 2 * k * k, 2 * i * j + 2 * w * k, 2 * i * k - 2 * w * j, 0,
        2 * i * j - 2 * w * k, 1 - 2 * i * i - 2 * k * k, 2 * j * k + 2 * w * i, 0,
        2 * i * k + 2 * w * j, 2 * j * k - 2 * w * i, 1 - 2 * i * i - 2 * j * j, 0,
        0, 0, 0, 1
    )
end

--- @brief Remove translation and scale, leaving only normalized rotation axes
function rt.Transform:set_to_orientation()
    local x11, x12, x13, _,
    x21, x22, x23, _,
    x31, x32, x33 = self._native:getMatrix()

    x11, x12, x13 = math.normalize(x11, x12, x13)
    x21, x22, x23 = math.normalize(x21, x22, x23)
    x31, x32, x33 = math.normalize(x31, x32, x33)

    self._native:setMatrix(
        x11, x12, x13, 0,
        x21, x22, x23, 0,
        x31, x32, x33, 0,
        0, 0, 0, 1
    )
end

--- @brief Get the underlying LÖVE transform for direct use
function rt.Transform:get_native()
    return self._native
end

File: website/game/src/common/transform.ts
import { Vec2, Vec3 } from "./vector.ts";

/** **/
export class Transform {
    private data: Float32Array;

    /** **/
    constructor(data?: Float32Array) {
        this.data = data ?? new Float32Array([
            1, 0, 0, 0,
            0, 1, 0, 0,
            0, 0, 1, 0,
            0, 0, 0, 1,
        ]);
    }

    /** **/
    public setData(data: Float32Array) : Transform {
        if (data.length != 4 * 4)
            throw new Error("In Transform.set: data does not have 16 entries")

        this.data = data;
        return this;
    }

    /** **/
    public getData() : Float32Array {
        return this.data;
    }

    /** **/
    public clone() : Transform {
        return new Transform(this.data)
    }

    /** **/
    public apply<T extends Vec2 | Vec3 | Transform>(other: T, write_to?: T): T {
        if (other instanceof Vec2)
            return this.multiply_vec2(other, write_to as Vec2) as T;
        else if (other instanceof Vec3)
            return this.multiply_vec3(other, write_to as Vec3) as T;
        else
            return this.multiply_transform(other as Transform, write_to as Transform) as T;
    }

    /** **/
    public translate(x: number, y: number, z: number = 0): Transform {
        const e = this.data;
        e[12] += e[0] * x + e[4] * y + e[8]  * z;
        e[13] += e[1] * x + e[5] * y + e[9]  * z;
        e[14] += e[2] * x + e[6] * y + e[10] * z;
        e[15] += e[3] * x + e[7] * y + e[11] * z;
        return this;
    }

    /** **/
    public scale(x: number, y?: number, z: number = 1): Transform {
        if (y === undefined) y = x;

        const e = this.data;
        e[0]  *= x; e[1]  *= x; e[2]  *= x; e[3]  *= x;
        e[4]  *= y; e[5]  *= y; e[6]  *= y; e[7]  *= y;
        e[8]  *= z; e[9]  *= z; e[10] *= z; e[11] *= z;
        return this;
    }

    /** **/
    public rotate(angle: number): Transform {
        const c = Math.cos(angle);
        const s = Math.sin(angle);
        const e = this.data;
        const column_0_x = e[0], column_0_y = e[1], column_0_z = e[2], column_0_w = e[3];
        const column_1_x = e[4], column_1_y = e[5], column_1_z = e[6], column_1_w = e[7];
        e[0] = c * column_0_x + s * column_1_x;
        e[1] = c * column_0_y + s * column_1_y;
        e[2] = c * column_0_z + s * column_1_z;
        e[3] = c * column_0_w + s * column_1_w;
        e[4] = -s * column_0_x + c * column_1_x;
        e[5] = -s * column_0_y + c * column_1_y;
        e[6] = -s * column_0_z + c * column_1_z;
        e[7] = -s * column_0_w + c * column_1_w;
        return this;
    }

    /** **/
    public transpose(write_to?: Transform): Transform {
        const e = this.data;
        const target = write_to ?? new Transform();
        target.data.set([
            e[0], e[4], e[8],  e[12],
            e[1], e[5], e[9],  e[13],
            e[2], e[6], e[10], e[14],
            e[3], e[7], e[11], e[15],
        ]);
        return target;
    }

    /** **/
    public asIdentity(): Transform {
        this.data.set([
            1, 0, 0, 0,
            0, 1, 0, 0,
            0, 0, 1, 0,
            0, 0, 0, 1
        ]);

        return this;
    }

    /** **/
    public asPerspectiveProjection(fov: number, aspect: number, near: number, far: number): Transform {
        const top    =  near * Math.tan(fov / 2);
        const bottom = -top;
        const right  =  top * aspect;
        const left   = -right;
        this.data.set([
            2 * near / (right - left), 0, 0, 0,
            0, 2 * near / (bottom - top), 0, 0,
            0, 0, -(far + near) / (far - near), -1,
            0, 0, -2 * far * near / (far - near), 0
        ]);
        return this;
    }

    /** **/
    public asOrthographicProjection(width: number, height: number, near: number, far: number): Transform {
        const right  =  width  / 2;
        const left   = -right;
        const top    = -height / 2;  // y-down: top is negative
        const bottom =  height / 2;
        this.data.set([
            2 / (right - left), 0, 0, 0,
            0, 2 / (top - bottom), 0, 0,
            0, 0, -2 / (far - near), 0,
            -((right + left) / (right - left)), -((top + bottom) / (top - bottom)), -((far + near) / (far - near)), 1,
        ]);
        return this;
    }

    /** **/
    public inverse(write_to?: Transform): Transform {
        const e = this.data;
        const target = (write_to ?? new Transform());
        const t = new Float32Array(16);

        t[0]  =  e[5]*e[10]*e[15] - e[5]*e[11]*e[14] - e[9]*e[6]*e[15]  + e[9]*e[7]*e[14]  + e[13]*e[6]*e[11]  - e[13]*e[7]*e[10];
        t[4]  = -e[4]*e[10]*e[15] + e[4]*e[11]*e[14] + e[8]*e[6]*e[15]  - e[8]*e[7]*e[14]  - e[12]*e[6]*e[11]  + e[12]*e[7]*e[10];
        t[8]  =  e[4]*e[9]*e[15]  - e[4]*e[11]*e[13] - e[8]*e[5]*e[15]  + e[8]*e[7]*e[13]  + e[12]*e[5]*e[11]  - e[12]*e[7]*e[9];
        t[12] = -e[4]*e[9]*e[14]  + e[4]*e[10]*e[13] + e[8]*e[5]*e[14]  - e[8]*e[6]*e[13]  - e[12]*e[5]*e[10]  + e[12]*e[6]*e[9];
        t[1]  = -e[1]*e[10]*e[15] + e[1]*e[11]*e[14] + e[9]*e[2]*e[15]  - e[9]*e[3]*e[14]  - e[13]*e[2]*e[11]  + e[13]*e[3]*e[10];
        t[5]  =  e[0]*e[10]*e[15] - e[0]*e[11]*e[14] - e[8]*e[2]*e[15]  + e[8]*e[3]*e[14]  + e[12]*e[2]*e[11]  - e[12]*e[3]*e[10];
        t[9]  = -e[0]*e[9]*e[15]  + e[0]*e[11]*e[13] + e[8]*e[1]*e[15]  - e[8]*e[3]*e[13]  - e[12]*e[1]*e[11]  + e[12]*e[3]*e[9];
        t[13] =  e[0]*e[9]*e[14]  - e[0]*e[10]*e[13] - e[8]*e[1]*e[14]  + e[8]*e[2]*e[13]  + e[12]*e[1]*e[10]  - e[12]*e[2]*e[9];
        t[2]  =  e[1]*e[6]*e[15]  - e[1]*e[7]*e[14]  - e[5]*e[2]*e[15]  + e[5]*e[3]*e[14]  + e[13]*e[2]*e[7]   - e[13]*e[3]*e[6];
        t[6]  = -e[0]*e[6]*e[15]  + e[0]*e[7]*e[14]  + e[4]*e[2]*e[15]  - e[4]*e[3]*e[14]  - e[12]*e[2]*e[7]   + e[12]*e[3]*e[6];
        t[10] =  e[0]*e[5]*e[15]  - e[0]*e[7]*e[13]  - e[4]*e[1]*e[15]  + e[4]*e[3]*e[13]  + e[12]*e[1]*e[7]   - e[12]*e[3]*e[5];
        t[14] = -e[0]*e[5]*e[14]  + e[0]*e[6]*e[13]  + e[4]*e[1]*e[14]  - e[4]*e[2]*e[13]  - e[12]*e[1]*e[6]   + e[12]*e[2]*e[5];
        t[3]  = -e[1]*e[6]*e[11]  + e[1]*e[7]*e[10]  + e[5]*e[2]*e[11]  - e[5]*e[3]*e[10]  - e[9]*e[2]*e[7]    + e[9]*e[3]*e[6];
        t[7]  =  e[0]*e[6]*e[11]  - e[0]*e[7]*e[10]  - e[4]*e[2]*e[11]  + e[4]*e[3]*e[10]  + e[8]*e[2]*e[7]    - e[8]*e[3]*e[6];
        t[11] = -e[0]*e[5]*e[11]  + e[0]*e[7]*e[9]   + e[4]*e[1]*e[11]  - e[4]*e[3]*e[9]   - e[8]*e[1]*e[7]    + e[8]*e[3]*e[5];
        t[15] =  e[0]*e[5]*e[10]  - e[0]*e[6]*e[9]   - e[4]*e[1]*e[10]  + e[4]*e[2]*e[9]   + e[8]*e[1]*e[6]    - e[8]*e[2]*e[5];

        const inverse_determinant = 1.0 / (e[0]*t[0] + e[1]*t[4] + e[2]*t[8] + e[3]*t[12]);
        for (let i = 0; i < 16; i++) t[i] *= inverse_determinant;

        target.setData(t);
        return target;
    }

    // ### internal ###

    private multiply_transform(other: Transform, write_to?: Transform): Transform {
        const a = this.data;
        const b = other.data;
        const t = new Float32Array(16);

        t[0]  = a[0]*b[0]  + a[4]*b[1]  + a[8]*b[2]  + a[12]*b[3];
        t[4]  = a[0]*b[4]  + a[4]*b[5]  + a[8]*b[6]  + a[12]*b[7];
        t[8]  = a[0]*b[8]  + a[4]*b[9]  + a[8]*b[10] + a[12]*b[11];
        t[12] = a[0]*b[12] + a[4]*b[13] + a[8]*b[14] + a[12]*b[15];
        t[1]  = a[1]*b[0]  + a[5]*b[1]  + a[9]*b[2]  + a[13]*b[3];
        t[5]  = a[1]*b[4]  + a[5]*b[5]  + a[9]*b[6]  + a[13]*b[7];
        t[9]  = a[1]*b[8]  + a[5]*b[9]  + a[9]*b[10] + a[13]*b[11];
        t[13] = a[1]*b[12] + a[5]*b[13] + a[9]*b[14] + a[13]*b[15];
        t[2]  = a[2]*b[0]  + a[6]*b[1]  + a[10]*b[2] + a[14]*b[3];
        t[6]  = a[2]*b[4]  + a[6]*b[5]  + a[10]*b[6] + a[14]*b[7];
        t[10] = a[2]*b[8]  + a[6]*b[9]  + a[10]*b[10]+ a[14]*b[11];
        t[14] = a[2]*b[12] + a[6]*b[13] + a[10]*b[14]+ a[14]*b[15];
        t[3]  = a[3]*b[0]  + a[7]*b[1]  + a[11]*b[2] + a[15]*b[3];
        t[7]  = a[3]*b[4]  + a[7]*b[5]  + a[11]*b[6] + a[15]*b[7];
        t[11] = a[3]*b[8]  + a[7]*b[9]  + a[11]*b[10]+ a[15]*b[11];
        t[15] = a[3]*b[12] + a[7]*b[13] + a[11]*b[14]+ a[15]*b[15];

        if (write_to === undefined)
            return new Transform(t);
        else
            return write_to.setData(t);
    }

    private multiply_vec2(vector: Vec2, write_to?: Vec2): Vec2 {
        const e = this.data;
        const target = write_to ?? vector;
        const previous_x = vector.x;
        const previous_y = vector.y;
        target.x = e[0] * previous_x + e[4] * previous_y + e[12];
        target.y = e[1] * previous_x + e[5] * previous_y + e[13];
        return target;
    }

    private multiply_vec3(vector : Vec3, write_to: Vec3): Vec3 {
        const e = this.data;
        const target = write_to ?? new Vec3();
        const previous_x = vector.x;
        const previous_y = vector.y;
        const previous_z = vector.z;

        target.x = e[0] * previous_x + e[4] * previous_y + e[8] * previous_z + e[12];
        target.y = e[1] * previous_x + e[5] * previous_y + e[9] * previous_z + e[13];
        target.z = e[2] * previous_x + e[6] * previous_y + e[10] * previous_z + e[14];

        // perspective divide
        const w = e[3] * previous_x + e[7] * previous_y + e[11] * previous_z + e[15];
        if (w !== 1.0 && w !== 0.0) {
            target.x /= w;
            target.y /= w;
            target.z /= w;
        }

        return target;
    }
}